File: CS/S170863/RowInplaceEditorHelper.cs
using DevExpress.Utils;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraVerticalGrid;
using DevExpress.XtraVerticalGrid.Events;
using DevExpress.XtraVerticalGrid.Rows;
using DevExpress.XtraVerticalGrid.ViewInfo;
using System.Drawing;
using System.Windows.Forms;

namespace WindowsFormsApplication1
{
    public class RowInplaceEditorHelper {
        private RepositoryItem _Item;
        private BaseRow _Row;
        private VGridControlBase _VGrid;

        private object _EditValue;
        public object EditValue {
            get { return _EditValue; }
            set { _EditValue = value; }
        }

        private BaseEdit _ActiveEditor;
        public BaseEdit ActiveEditor {
            get { return _ActiveEditor; }
            set { _ActiveEditor = value; }
        }

        public RowInplaceEditorHelper(BaseRow row, RepositoryItem inplaceEditor) {
            _Row = row;
            _Item = inplaceEditor;
            _VGrid = _Row.Grid;
            _VGrid.CustomDrawRowHeaderCell += vGrid_CustomDrawRowHeaderCell;
            _VGrid.MouseDown += vGrid_MouseDown;
        }

        void vGrid_CustomDrawRowHeaderCell(object sender, CustomDrawRowHeaderCellEventArgs e) {
            if(e.Row == _Row) {
                e.Caption = string.Empty;
                e.Handled = true;
                DrawEditorHelper.DrawRowInplaceEditor(e, _Item, EditValue);
            }
        }

        void vGrid_MouseDown(object sender, MouseEventArgs e) {
            CloseEditor();
            Rectangle editorBounds;
            if(ClickInEditor(e, out editorBounds)) {
                ShowEditor(editorBounds);
                DXMouseEventArgs.GetMouseArgs(e).Handled = true;
            }
        }

        private bool ClickInEditor(MouseEventArgs e, out Rectangle editorBounds) {
            editorBounds = Rectangle.Empty;
            Point location = e.Location;
            VGridHitInfo hitInfo = _VGrid.CalcHitInfo(location);
            BaseRowViewInfo viewInfo = _VGrid.ViewInfo[location];
            if(hitInfo.HitInfoType == HitInfoTypeEnum.HeaderCell && _Row == viewInfo.Row) {
                editorBounds = viewInfo.HeaderInfo.HeaderCellsRect;
                return editorBounds.Contains(location);
            }
            return false;
        }

        private void ShowEditor(Rectangle bounds) {
            ActiveEditor = _Item.CreateEditor();
            ActiveEditor.Properties.LockEvents();
            ActiveEditor.Parent = _VGrid;
            ActiveEditor.Properties.Assign(_Item);
            ActiveEditor.Properties.AutoHeight = false;
            ActiveEditor.BackColor = _VGrid.ViewInfo.PaintAppearance.FocusedRow.BackColor;
            ActiveEditor.Location = bounds.Location;
            ActiveEditor.Size = bounds.Size;
            ActiveEditor.CreateControl();
            ActiveEditor.EditValue = EditValue;
            ActiveEditor.SendMouse(ActiveEditor.PointToClient(Control.MousePosition), Control.MouseButtons);
            ActiveEditor.Properties.UnLockEvents();
        }

        private void CloseEditor() {
            if(ActiveEditor != null) {
                EditValue = ActiveEditor.EditValue;
                ActiveEditor.Dispose();
                ActiveEditor = null;
            }
        }
    }
}

File: VB/S170863/RowInplaceEditorHelper.vb
Imports DevExpress.Utils
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraVerticalGrid
Imports DevExpress.XtraVerticalGrid.Events
Imports DevExpress.XtraVerticalGrid.Rows
Imports DevExpress.XtraVerticalGrid.ViewInfo

Namespace WindowsFormsApplication1
    Public Class RowInplaceEditorHelper
        Private _Item As RepositoryItem
        Private _Row As BaseRow
        Private _VGrid As VGridControlBase

        Private _EditValue As Object
        Public Property EditValue() As Object
            Get
                Return _EditValue
            End Get
            Set(ByVal value As Object)
                _EditValue = value
            End Set
        End Property

        Private _ActiveEditor As BaseEdit
        Public Property ActiveEditor() As BaseEdit
            Get
                Return _ActiveEditor
            End Get
            Set(ByVal value As BaseEdit)
                _ActiveEditor = value
            End Set
        End Property

        Public Sub New(ByVal row As BaseRow, ByVal inplaceEditor As RepositoryItem)
            _Row = row
            _Item = inplaceEditor
            _VGrid = _Row.Grid
            AddHandler _VGrid.CustomDrawRowHeaderCell, AddressOf vGrid_CustomDrawRowHeaderCell
            AddHandler _VGrid.MouseDown, AddressOf vGrid_MouseDown
        End Sub

        Private Sub vGrid_CustomDrawRowHeaderCell(ByVal sender As Object, ByVal e As CustomDrawRowHeaderCellEventArgs)
            If e.Row Is _Row Then
                e.Caption = String.Empty
                e.Handled = True
                DrawEditorHelper.DrawRowInplaceEditor(e, _Item, EditValue)
            End If
        End Sub

        Private Sub vGrid_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs)
            CloseEditor()
            Dim editorBounds As Rectangle = Nothing
            If ClickInEditor(e, editorBounds) Then
                ShowEditor(editorBounds)
                DXMouseEventArgs.GetMouseArgs(e).Handled = True
            End If
        End Sub

        Private Function ClickInEditor(ByVal e As MouseEventArgs, ByRef editorBounds As Rectangle) As Boolean
            editorBounds = Rectangle.Empty
            Dim location As Point = e.Location
            Dim hitInfo As VGridHitInfo = _VGrid.CalcHitInfo(location)
            Dim viewInfo As BaseRowViewInfo = _VGrid.ViewInfo(location)
            If hitInfo.HitInfoType = HitInfoTypeEnum.HeaderCell AndAlso _Row Is viewInfo.Row Then
                editorBounds = viewInfo.HeaderInfo.HeaderCellsRect
                Return editorBounds.Contains(location)
            End If
            Return False
        End Function

        Private Sub ShowEditor(ByVal bounds As Rectangle)
            ActiveEditor = _Item.CreateEditor()
            ActiveEditor.Properties.LockEvents()
            ActiveEditor.Parent = _VGrid
            ActiveEditor.Properties.Assign(_Item)
            ActiveEditor.Properties.AutoHeight = False
            ActiveEditor.BackColor = _VGrid.ViewInfo.PaintAppearance.FocusedRow.BackColor
            ActiveEditor.Location = bounds.Location
            ActiveEditor.Size = bounds.Size
            ActiveEditor.CreateControl()
            ActiveEditor.EditValue = EditValue
            ActiveEditor.SendMouse(ActiveEditor.PointToClient(Control.MousePosition), Control.MouseButtons)
            ActiveEditor.Properties.UnLockEvents()
        End Sub

        Private Sub CloseEditor()
            If ActiveEditor IsNot Nothing Then
                EditValue = ActiveEditor.EditValue
                ActiveEditor.Dispose()
                ActiveEditor = Nothing
            End If
        End Sub
    End Class
End Namespace
